src/mito2/src/region/opener.rs

```diff
@@ -16,41 +16,50 @@
 
 use std::sync::Arc;
 
+use common_query::Output;
+use common_telemetry::info;
+use futures::StreamExt;
 use object_store::util::join_dir;
 use object_store::ObjectStore;
-use snafu::{ensure, OptionExt};
+use snafu::{ensure, OptionExt, ResultExt};
+use store_api::logstore::LogStore;
 use store_api::metadata::RegionMetadata;
-use store_api::storage::RegionId;
+use store_api::storage::{RegionId, SequenceNumber};
 
 use crate::config::MitoConfig;
-use crate::error::{RegionCorruptedSnafu, RegionNotFoundSnafu, Result};
+use crate::error::{RecvSnafu, RegionCorruptedSnafu, RegionNotFoundSnafu, Result};
 use crate::manifest::manager::{RegionManifestManager, RegionManifestOptions};
 use crate::memtable::MemtableBuilderRef;
-use crate::region::version::{VersionBuilder, VersionControl};
+use crate::region::version::{VersionBuilder, VersionControl, VersionControlRef};
 use crate::region::MitoRegion;
+use crate::region_writer::RegionWriteCtx;
+use crate::wal::Wal;
 
 /// Builder to create a new [MitoRegion] or open an existing one.
-pub(crate) struct RegionOpener {
+pub(crate) struct RegionOpener<S> {
     region_id: RegionId,
     metadata: Option<RegionMetadata>,
     memtable_builder: MemtableBuilderRef,
     object_store: ObjectStore,
     region_dir: String,
+    wal: Wal<S>,
```

Comment:
question: We only need wal when opening the region, could we pass it in `RegionOpener::open()`?
```rust
pub(crate) async fn open<S: LogStore>(self, config: &MitoConfig, wal: Wal<S>) -> Result<MitoRegion> {}
```